drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1082 @@
+	} else {
+		fprintf(stderr, "exiting\n");
+	}
+	return ret;
+}
+
+/*
+ * modeset_page_flip_event() is a callback-helper for modeset_draw() below.
+ * Please see modeset_draw() for more information.
+ *
+ * Note that this does nothing if the device is currently cleaned up. This
+ * allows to wait for outstanding page-flips during cleanup.
+ */
+
+static void modeset_page_flip_event(int fd, unsigned int frame,
+				    unsigned int sec, unsigned int usec,
+				    void *data)
+{
+	struct modeset_dev *dev = data;
+
+	dev->pflip_pending = false;
+	if (!dev->cleanup)
+		modeset_draw_dev(fd, dev);
+}
+
+static int set_crtc_property(drmModeAtomicReq *req,
+	struct drm_object *crtc, uint32_t crtc_id,
+	const char *name, uint64_t value)
+{
+	int i, prop_id = -1;
+
+	for (i = 0; i < crtc->props->count_props; i++) {
+		if (!strcmp(crtc->props_info[i]->name, name)) {
+			prop_id = crtc->props_info[i]->prop_id;
+			break;
+		}
+	}
+
+	if (prop_id < 0) {
+                fprintf(stderr, "no crtc property: %s\n", name);
+		return -EINVAL;
+	}
+
+	return drmModeAtomicAddProperty(req, crtc_id, prop_id, value);
+}
+
+static int set_connector_property(drmModeAtomicReq *req,
+	struct drm_object *connector, uint32_t conn_id,
+	const char *name, uint64_t value)
+{
+	int i, prop_id = -1;
+
+	for (i = 0; i < connector->props->count_props; i++) {
+		if (!strcmp(connector->props_info[i]->name, name)) {
+			prop_id = connector->props_info[i]->prop_id;
+			break;
+		}
+	}
+
+	if (prop_id < 0) {
+                fprintf(stderr, "no connector property: %s\n", name);
+		return -EINVAL;
+	}
+
+	return drmModeAtomicAddProperty(req, conn_id, prop_id, value);
+}
+
+static int set_plane_property(drmModeAtomicReq *req,
+	struct drm_object *plane, uint32_t plane_id,
+	const char *name, uint64_t value)
+{
+        int i, prop_id = -1;
+
+        for (i = 0 ; i < plane->props->count_props ; i++) {
+                if (!strcmp(plane->props_info[i]->name, name)) {
+                        prop_id = plane->props_info[i]->prop_id;
+                        break;
+                }
+        }
+
+        if (prop_id < 0) {
+                fprintf(stderr, "no plane property: %s\n", name);
+                return -EINVAL;
+        }
+
+        return drmModeAtomicAddProperty(req, plane_id, prop_id, value);
+}
+
+static int modeset_atomic_modeset(int fd, struct modeset_dev *dev)
+{
+	drmModeAtomicReq *req;
+	uint32_t blob_id;
+	int ret, flags;
+
+	req = drmModeAtomicAlloc();
+
+	if (set_connector_property(req, &dev->connector, dev->conn_id, "CRTC_ID", dev->crtc_id) < 0)
+		return -1;
+
+	if (drmModeCreatePropertyBlob(fd, &dev->mode, sizeof(dev->mode), &blob_id) != 0)
+		return -1;
+
+	if (set_crtc_property(req, &dev->crtc, dev->crtc_id, "MODE_ID", blob_id) < 0)
+		return -1;
+
+	if (set_crtc_property(req, &dev->crtc, dev->crtc_id, "ACTIVE", 1) < 0)
+		return -1;
+
+	flags = DRM_MODE_ATOMIC_ALLOW_MODESET;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+
+	drmModeAtomicFree(req);
```

Comment:
I would not expect a modeset without a framebuffer to work. This may work by accident, if the CRTC is already active and has a framebuffer.

Reply:
Yes, this is absolutely required.

Reply:
> I would not expect a modeset without a framebuffer to work.

Just to be sure: I should have done something similar to [modeset_atomic_commit()](https://github.com/dvdhrm/docs/pull/10/files#diff-f78817f25c5ec8cefcd9fa42fbbd6cd0R833), where we set the CRTC and fb properties for the plane, right?

Reply:
I guess it finally clicked hehe. That's why the atomic commit was not working without `DRM_FBDEV_EMULATION=y` . While using the legacy fbdev support for the KMS driver, the CRTC was already active and had a framebuffer.